design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,241 @@
+  been deprecated [has been removed in
+  Queens](https://docs.openstack.org/keystone/latest/contributor/http-api.html)
+  which was release in 2018 and is now in extended maintenance mode
+  (which means it gets no more points releases and only sporadic bug
+  fixes/security fixes).
+
+* Authenticating Ceph RGW to Keystone via admin token – Only
+  authentication via an OpenStack service account will be supported.
+
+* Support for APIs beside S3 and Swift.
+
+* Interaction of Swift with OBCs is out of scope of this document.  If
+  you need to access a bucket created by OBC via Swift you need to
+  create a separate `cephobjectstoreuser`, configure its access rights
+  to the bucket and use those credentials.
+
+* Support for authentication technologies other than Keystone (e.g. LDAP)
+
+* Exposing options that disable security features (e.g. TLS verification)
+
+## Proposal details
+
+The Object Store CRD will have to be extended to accommodate the new
+settings.
+
+### Keystone integration
+
+A new section `auth:` is added to the Object Store CRD. To configure
+the keystone integration:
+
+```yaml
+auth:
+  keystone:
+    url: https://keystone:5000/
+    acceptedRoles: ["_member_", "service", "admin"]
+    apiVersion: 3
+    implicitTenants: swift
+    tokenCacheSize: 1000
+    revocationInterval: 1200
+    serviceUserSecret: rgw-service-user
+```
+
+The certificate to verify the Keystone endpoint can't be explicitly
+configured in Ceph RGW, the system configuration of the pod is used,
+you can add to the system certificate store via the
+`gateway.caBundleRef` setting of the object store CRD.
+
+The credentials for the Keystone service account used by Ceph RGW are
+supplied in a Secret that contains a mapping of OpenStack openrc
+environment variables. Only password authentication to Keystone is
+supported. Example:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: rgw-service-user
+data:
+  OS_PASSWORD: "horse staples battery correct"
+  OS_USERNAME: "ceph-rgw"
+  OS_PROJECT_NAME: "admin"
+  OS_USER_DOMAIN_NAME: "Default"
+  OS_PROJECT_DOMAIN_NAME: "Default"
+  OS_AUTH_TYPE: "password"
+```
+
+This format is chosen because it is a natural and interoperable way
+that keystone credentials are represented and for the supported auth
+type it maps naturally to the Ceph RGW configuration.
+
+The following constraints must be fulfilled by the secret:
+* `OS_AUTH_TYPE` must be `password` or omitted.
+* `OS_USER_DOMAIN_NAME` must equal `OS_PROJECT_DOMAIN_NAME`.
+* All openrc variables not in the example above (e.g. API version and
+  Keystone endpoint) are ignored.
+
+### Swift integration
+
+Swift is configured in the `gateway:` section of the Object Store CRD.
+
+The currently unused `type:` argument may now take the values `swift`
+or `s3`.
+
+The following new settings are available to properly configure Swift:
+```yaml
+gateway:
+  type: swift
+  swiftAccountInUrl: true
+  swiftUrlPrefix: /swifter
+  swiftVersioningEnabled: false
```

Comment:
I have generally seen that it isn't recommended to use configurations that look like 
```
type: 'thing'
thingSpecificConfig: blah
```

This means that the configs are tied specifically to the value in `type`. Instead, having a different struct for a different type able to specify what is optional and required is preferable. LIke this:

```
protocol:
  swift:
    accountInURL: t/f
    urlPrefix: "/example"
    versioningEnabled: t/f
```

Since we currently ignore `type`, lets just continue doing so. For backwards compatibility, we should keep the existing configs and assume `s3`, unless `protocol` is configured.

Reply:
This matches with one of my proposals for supporting swift and s3 in a single RGW instance. I've adapted the design document accordingly.